### ISSUE TITLE
New string literals are marked as tainted if previously equal tainted strings were read from a file

### ISSUE THREAD

**Issue by rsim** [jruby/jruby#7018]
**Environment Information**
  
Provide at least:
* JRuby version (`jruby -v`) and command line (flags, JRUBY_OPTS, etc)
   `jruby 9.3.3.0 (2.6.8) 2022-01-19 b26de1f5c5 OpenJDK 64-Bit Server VM 11.0.13+8-LTS on 11.0.13+8-LTS [darwin-aarch64]`
* Operating system and platform (e.g. `uname -a`)
   `Darwin Kernel Version 21.2.0: Sun Nov 28 20:28:41 PST 2021; root:xnu-8019.61.5~1/RELEASE_ARM64_T6000 arm64`

Other relevant info you may wish to add:
* Installed or activated gems
   `mime-types`

**Expected Behavior**

`mime-types` gem loads from the `mime.content_type.column` many different MIME content types. It contains, for example, one line
```
application/java-archive jar
```
When running on MRI 2.5.8 and loading mime-types and then creating new strings (which are equal to the previous strings created from the loaded file) then they are not tainted:
```
$ gem install mime-types
$ ruby -e 'gem "mime-types"; require "mime/types"; puts "jar".tainted?, "application/java-archive".tainted?'
false
false
```

**Actual Behavior**

When running this on JRuby 9.3.3.0 or 9.3.2.0 then the newly created strings are tainted (which is wrong):
```
$ gem install mime-types
$ ruby -e 'gem "mime-types"; require "mime/types"; puts "jar".tainted?, "application/java-archive".tainted?'
true
true
```

When running this on JRuby 9.2.20.1 then some new strings are not tainted but some are tainted:
```
$ gem install mime-types
$ ruby -e 'gem "mime-types"; require "mime/types"; puts "jar".tainted?, "application/java-archive".tainted?'
false
true
```

If equal strings are created before requiring `mime-types` then all string are correctly untainted:
```
$ ruby -e 'gem "mime-types"; puts "jar".tainted?, "application/java-archive".tainted?; require "mime/types"; puts "jar".tainted?, "application/java-archive".tainted?'
false
false
false
false
```

It seems that there is a problem that for newly created strings the previously created tainted strings are reused.
I couldn't replicate this bug with a smaller example, it seems that it happens only with larger files and more string objects as in this case.

The side effect of this bug was that `Gem::BasicSpecification#have_file?` was failing for one gem in our application as in this method the following line was failing
```
base = File.join(gems_dir, full_name, path.tap(&Gem::UNTAINT), file).tap(&Gem::UNTAINT)
```
as `path` was `"jar"` which due to this bug was tainted string (and frozen) and therefore `untaint` of it failed with FrozenError.

**Comment by headius** [comment]
Huh, well I never expected to get a tainting bug this late in the game.

We make no guarantees about tainting in JRuby and only try to do it to match visible behavior. Tainting is deprecated in 2.7 and gone in 3.0 and we have already removed it on master.

This should be simple enough to fix but it will probably just be us removing the tainting altogether.

**Comment by rsim** [comment]
As we don't really need this tainting in our application then I'm wondering what would be the easiest way to patch tainting methods so that they don't do anything?

Currently I added `Gem::UNTAINT = proc {}` to avoid FrozenError in `path.tap(&Gem::UNTAINT)` (which is the behavior of  Gem::UNTAINT starting from Ruby 2.7).

**Comment by headius** [comment]
This seems to be specific to the interpreter:

```
$ ruby -v -e 'gem "mime-types"; require "mime/types"; puts "jar".tainted?, "application/java-archive".tainted?'
jruby 9.3.2.0 (2.6.8) 2021-12-01 0b8223f905 OpenJDK 64-Bit Server VM 16-ea+29-2091 on 16-ea+29-2091 +jit [darwin-x86_64]
false
false

$ ruby --dev -v -e 'gem "mime-types"; require "mime/types"; puts "jar".tainted?, "application/java-archive".tainted?'
jruby 9.3.2.0 (2.6.8) 2021-12-01 0b8223f905 OpenJDK 64-Bit Server VM 16-ea+29-2091 on 16-ea+29-2091 [darwin-x86_64]
true
true
```

I can reproduce on 9.3.2 and 9.3.3.

**Comment by headius** [comment]
CRuby avoids deduplicating string keys that are tainted:

```c
VALUE
rb_hash_key_str(VALUE key)
{
    if (!RB_FL_ANY_RAW(key, FL_TAINT|FL_EXIVAR) && RBASIC_CLASS(key) == rb_cString) {
        return rb_fstring(key);
    }
    else {
        return rb_str_new_frozen(key);
    }
}
```

Fixing this in JRuby gets 9.3 back to 9.2 behavior. Still looking for the source of the second tainting.

**Comment by headius** [comment]
The second deduplication of "application/java-archive" happens within mime-types types.rb:

```ruby
  def content_type=(type_string)
    match = MEDIA_TYPE_RE.match(type_string)
    fail InvalidContentType, type_string if match.nil?

    @content_type = intern_string(type_string)
    @raw_media_type, @raw_sub_type = match.captures
    @simplified = intern_string(MIME::Type.simplified(match))
    @i18n_key = intern_string(MIME::Type.i18n_key(match))
    @media_type, @sub_type = MEDIA_TYPE_RE.match(@simplified).captures

    @raw_media_type = intern_string(@raw_media_type)
    @raw_sub_type = intern_string(@raw_sub_type)
    @media_type = intern_string(@media_type)
    @sub_type = intern_string(@sub_type)
  end

  if String.method_defined?(:-@)
    def intern_string(string)
      -string
    end
  else
    # MRI 2.2 and older don't have a method for string interning,
    # so we simply freeze them for keeping a similar interface
    def intern_string(string)
      string.freeze
    end
  end
```

But despite CRuby not checking tainting in the `String#-@` internment method, they do not reproduce this issue. I need to confirm whether the strings is tainted at the point of calling `-@` in CRuby.

**Comment by headius** [comment]
CRuby has a second check for tainting inside the deduplication rb_fstring function, via the BARE_STRING_P check.

```c
#define BARE_STRING_P(str) (!FL_ANY_RAW(str, FL_TAINT|FL_EXIVAR) && RBASIC_CLASS(str) == rb_cString)
...
rb_fstring(VALUE str)
{
    VALUE fstr;
    int bare;

    Check_Type(str, T_STRING);

    if (FL_TEST(str, RSTRING_FSTR))
        return str;

    bare = BARE_STRING_P(str);
    if (!bare) {
        if (STR_EMBED_P(str)) {
            OBJ_FREEZE_RAW(str);
            return str;
        }
        if (FL_TEST_RAW(str, STR_NOEMBED|STR_SHARED_ROOT|STR_SHARED) == (STR_NOEMBED|STR_SHARED_ROOT)) {
            assert(OBJ_FROZEN(str));
            return str;
        }
    }
...
```

I will add an equivalent check.

**Comment by headius** [comment]
Fixes are landing for 9.2 and 9.3 now.

Tainted strings should not be deduplicated by `String#-@` or `Hash` string key logic. The fixes in #7020 and 7021 will avoid this by checking taint in the same places as CRuby.

This appeared to be an interpreter issue because the interpreter constructs new mutable strings by duping a frozen deduplicated version of the same string (same mechanism as CRuby). When the cache was polluted, this meant all new mutable strings with the same bytes were infected by taint. The JIT does not dup a frozen string and always creates a clean new String object, hiding the cache tainting bug that was still present.

Basically, the pollution of the string deduplication cache was the root issue, exposed more immediately by the way the interpreter constructs mutable strings.

We are discussing what, if anything, should change in JIT vs interpreter regarding this last discovery. With an unpolluted cache, there's no visible difference between the two.

**Comment by headius** [comment]
Fixed in 9.2.21.0 and 9.3.4.0.

**Comment by headius** [comment]
I'm not sure if there's any great workarounds here. The elegant logic in the interpreter unfortunately means that tainted cached strings infect any mutable literal string with the same bytes, and there's no easy way to avoid that. The JIT will produce mutable strings without infection, but the underlying pollution of the cache is still present.
